compiler/rustc_builtin_macros/src/cfg_match.rs

```diff
@@ -0,0 +1,284 @@
+        [first, rest @ .., last] => {
+            let mut offset = first.1;
+            let mut prev_offset = 0;
+            cb(cx, no, &first.0, items.get_mut(prev_offset..offset).unwrap_or_default());
+            for elem in rest {
+                prev_offset = offset;
+                offset = elem.1;
+                cb(cx, no, &elem.0, items.get_mut(prev_offset..offset).unwrap_or_default());
+            }
+            prev_offset = offset;
+            offset = last.1;
+            cb_last(cx, no, &last.0, items.get_mut(prev_offset..offset).unwrap_or_default());
+        }
+    }
+    Ok(())
+}
+
+// #[cfg(all(** YES **, not(any(** NO **, ** NO **, ..))))]
+fn mk_cfg(
+    cx: &mut ExtCtxt<'_>,
+    no: &[&TokenStream],
+    sp: Span,
+    yes: Option<&TokenStream>,
+) -> Attribute {
+    let mut any_tokens = TokenStream::new(Vec::with_capacity(4));
+    if let &[first, ref rest @ ..] = no {
+        any_tokens.push_stream(first.clone());
+        for elem in rest.iter().copied() {
+            any_tokens.push_tree(TokenTree::token_alone(token::Comma, sp));
+            any_tokens.push_stream(elem.clone());
+        }
+    }
+
+    let mut not_tokens = TokenStream::new(Vec::with_capacity(2));
+    not_tokens.push_tree(TokenTree::token_alone(token::Ident(sym::any, false), sp));
+    not_tokens.push_tree(TokenTree::Delimited(
+        DelimSpan::from_single(sp),
+        token::Delimiter::Parenthesis,
+        any_tokens,
+    ));
+
+    let mut all_tokens = TokenStream::new(Vec::with_capacity(4));
+    if let Some(elem) = yes {
+        all_tokens.push_stream(elem.clone());
+        all_tokens.push_tree(TokenTree::token_alone(token::Comma, sp));
+    }
+    all_tokens.push_tree(TokenTree::token_alone(token::Ident(sym::not, false), sp));
+    all_tokens.push_tree(TokenTree::Delimited(
+        DelimSpan::from_single(sp),
+        token::Delimiter::Parenthesis,
+        not_tokens,
+    ));
+
+    let mut tokens = TokenStream::new(Vec::with_capacity(2));
+    tokens.push_tree(TokenTree::token_alone(token::Ident(sym::all, false), sp));
+    tokens.push_tree(TokenTree::Delimited(
+        DelimSpan::from_single(sp),
+        token::Delimiter::Parenthesis,
+        all_tokens,
+    ));
+
+    mk_attr(
+        &cx.sess.parse_sess.attr_id_generator,
+        AttrStyle::Outer,
+        Path::from_ident(Ident::new(sym::cfg, sp)),
+        AttrArgs::Delimited(DelimArgs {
+            dspan: DelimSpan::from_single(sp),
+            delim: token::Delimiter::Parenthesis,
+            tokens,
+        }),
+        sp,
+    )
+}
+
+fn parse<'session>(
+    cx: &mut ExtCtxt<'session>,
+    tts: TokenStream,
+) -> PResult<'session, (bool, Vec<P<Item>>, Vec<(TokenStream, usize)>)> {
+    let mut parser = cx.new_parser_from_tts(tts);
+    let mut does_not_have_wildcard = true;
+    let mut items = Vec::with_capacity(4);
+    let mut items_offsets = Vec::with_capacity(4);
```

Comment:
This auxiliary `items_offsets` vector is used to avoid creating `Vec<Vec<P<Item>>>`, i.e., all items are stored in the same contiguous `items` memory.

```
`items`: [ Items of first match, Items of second match, .. ]
          |______0 to 10_______| |______10 to 25______|    
                   |      _________________/
                   |     /
`items_offsets`: [ 10, 25, ..]
```